### DISCUSSION TITLE
Callback in Unity not working

### DISCUSSION THREAD

**Discussion by dharry1968** [jeffreylanters/react-unity-webgl#256] [Help Wanted!] [answered]
Hi all, i'm having an issue getting React to send a message back to Unity for some reason... I am using Unity 2019 and v7 of the library. 

The call from Unity to React works just fine. My React code is:

```js
async function server_login() {
  try {
    await ual.showModal();
    console.log('User Account:' + ual.users[0].accountName);
    console.log('Calling WalletController');
    unityContent.send(
      "WalletController",
      "Server_Response_LoginData",
      ual.users[0]?.accountName
    );
  } catch (error) {
    console.log('server_login Error:' + error);
  }
}
```
  
  and on the Unity side I have:
  
    public void Server_Response_LoginData(string account)
    {
        Debug.Log("Call back from Login: " + account);
        isLoginProcessRunning = false;
        SetLastSucessLoginWallet(account);
    }
    
Everything up to the console.log('Calling WalletController') works fine, but nothing seems to come back to Unity. What might I be doing wrong?

Also, why does some documentation refer to UnityContext and others refer to UnityContent? I tried switching to UnityContext but that wouldn't compile.

Regards
Dean

**Comment by jeffreylanters** [comment, marked as answer]
Hi. Ar you sure `ual.users[0]?.accountName` returns an actual string? Because it it returns undefined due to the optional chaining, the module handles it as if there was no parameter which means `public void Server_Response_LoginData()` would be invoked rather than `public void Server_Response_LoginData(string account)` with a string parameter.

**Reply by dharry1968** [reply]
yes definitely, even if I set it to a string manually I don't get the return, the console.log above for the account name does display a result.

![image](https://user-images.githubusercontent.com/790458/141080759-417c774f-09d5-4128-83bb-56dcfcb06cfb.png)

These are the results from the console.

**Reply by jeffreylanters** [reply]
Thanks. Just checking, is this script attached to a game object named exactly `WalletController`?

**Reply by dharry1968** [reply]
yes, exactly, iv'e checked that 10 times :)

**Reply by jeffreylanters** [reply]
Haha great. Something else that might be going on is that the UnityContent is detached from the UnityInstance, where / how have your declared your UnityContent? If, for example, its declared within a functional component without the use of a useEffect (componentDidMount), a rerender could cause a detach.

**Reply by dharry1968** [reply]
UnityContent is declared like this:  const unityContent = new UnityContent("Build.json", "UnityLoader.js"); it's in my App.js which is the same function that the call back is in.

Here's a pastebin if the whole function for reference: https://pastebin.com/1EZ8ZKU3

**Reply by jeffreylanters** [reply]
Could you try moving the `const unityContent = new...` line outside of the App.js? Just above the `function App()...` line

**Reply by dharry1968** [reply]
Ok, that works!! wow, something so simple! I would never have picked that - thank you so much!

**Reply by jeffreylanters** [reply]
Great! (Could you mark this thread as the Answer?)

Also, a quick explanation on why this happens. Since the `new UnityContent` is within the render method, and is not using any kind of state or life cycle methods, it's being recreated each time the component rerenders, resulting in the Unity Content losing its reference to the Unity Instance. Simply placing it outside of your render function or wrapping it within a "component did mount" cycle resolves this issue.

Best of luck with your project!

**Comment by jeffreylanters** [comment]
Oh and about your Content/Context question, this has been refactored in version 8.x. You should refer to the 7.x documentation instead of you are using Unity 2019. You can find it [here](https://github.com/jeffreylanters/react-unity-webgl/tree/7.1.13).

**Reply by dharry1968** [reply]
thanks, I guess that was the case, I followed through on this tutorial: https://keithweaver.ca/lessons/running-unity-webgl-inside-a-react-app/
